core/src/apps/cardano/sign_tx.py

```diff
@@ -4,238 +4,195 @@
 
 
-class Transaction:
-    def __init__(
-        self,
-        inputs: list,
-        outputs: list,
-        keychain,
-        protocol_magic: int,
-        input_coins_sum: int,
-    ):
-        self.inputs = inputs
-        self.outputs = outputs
-        self.keychain = keychain
-        # attributes have to be always empty in current Cardano
-        self.attributes = {}
-
-        self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic, "Unknown")
-        self.protocol_magic = protocol_magic
-        self.input_coins_sum = input_coins_sum
-
-    def _process_outputs(self):
-        change_addresses = []
-        change_derivation_paths = []
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
-
-        for output in self.outputs:
-            if output.address_n:
-                address, _ = derive_address_and_node(self.keychain, output.address_n)
-                change_addresses.append(address)
-                change_derivation_paths.append(output.address_n)
-                change_coins.append(output.amount)
-            else:
-                if output.address is None:
-                    raise wire.ProcessError(
-                        "Each output must have address or address_n field!"
-                    )
-                if not is_safe_output_address(output.address):
-                    raise wire.ProcessError("Invalid output address!")
-
-                outgoing_coins.append(output.amount)
-                output_addresses.append(output.address)
-
-        self.change_addresses = change_addresses
-        self.output_addresses = output_addresses
-        self.outgoing_coins = outgoing_coins
-        self.change_coins = change_coins
-        self.change_derivation_paths = change_derivation_paths
-
-    def _build_witnesses(self, tx_aux_hash: str):
-        witnesses = []
-        for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
-            signature = ed25519.sign_ext(
-                node.private_key(), node.private_key_ext(), message
-            )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
+def _serialize_tx(keychain: seed.Keychain, msg: CardanoSignTx) -> Tuple[bytes, bytes]:
+    tx_body = _build_tx_body(keychain, msg)
+    tx_hash = _hash_tx_body(tx_body)
 
-        return witnesses
+    witnesses_for_cbor = _build_witnesses(
+        keychain, msg.inputs, tx_hash, msg.protocol_magic
+    )
+    # byron witnesses have the key 2 in Shelley
+    witnesses = {2: _detupleize(witnesses_for_cbor)}
```

Comment:
This was done purely so that the types can be set correctly and therefore readability can be improved. 

We also thought about implementing it in CBOR although then there could be an ambiguity in decoding arrays - do we decode as list or a tuple?

Reply:
Just decode as array always. I'd file this under "API convenience": CBOR only recognizes arrays (right?), and so only returns arrays. But for usage convenience it will transparently encode tuples as arrays as well, because that's how we use it

Reply:
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/0ead85323b19db525f48336936246602f066f8eb). I'll change the commit message later, if we're not squashing to one commit.